zbus_macros/src/lib.rs

```diff
@@ -215,6 +215,15 @@ pub fn dbus_proxy(attr: TokenStream, item: TokenStream) -> TokenStream {
 /// properties or signal depending on the item attributes. It will implement the [`Interface`] trait
 /// `for T` on your behalf, to handle the message dispatching and introspection support.
 ///
+/// The trait accepts the `interface` attributes:
+///
+/// * `name` - the D-Bus interface name
+///
+/// * `no_spawn` - Do not spawn async tasks to call the methods. By default, zbus will spawn a task
+///   for each method call. This may result in methods being handled out of order and may not be the
+///   expected or desired behaviour. Note that D-Bus has currently not stated explicitely message
+///   ordering guarantees (see also <https://gitlab.freedesktop.org/dbus/dbus/-/issues/179>).
```

Comment:
* Remove "currently" and update to reflect that we're talking about method handling order. Message ordering ordering and this linked issue is very much irrelevant.
* I don't think this should be decided on per interfaces bases. I just want a simple boolean setting on either ConnectionBuilder or ObjectServer.

Reply:
@elmarco actually I've a better suggestion that you'll probably like: we can make non-spawning the default for p2p. Since separate clients will be on separate connections anyway and hence have their ObjectServer running in a separate task. 

Also we don't need to provide an option to change this behaviour unless someone asks for it. I don't see much point in parallelising method calls from the same client. We should document this behaviour though.

I'd be fine with doing the same for the bus case but that would be adding a lot of complexity to track all the peers and managing tasks for them.

Reply:
> @elmarco actually I've a better suggestion that you'll probably like: we can make non-spawning the default for p2p. Since separate clients will be on separate connections anyway and hence have their ObjectServer running in a separate task.

We could discuss the various use cases having an option to set the spawning behaviour at either Method, Interface, ObjectServer, Connection. They are *all* valid to me.  

But I think the sweet spot is at the Interface level, because it is a set of related methods and implementation, and you can document this expected behaviour better than at the ObjectServer or Connection level. At the ObjectServer or Connection level, you enforce a behaviour on other implementations that may not require it.

There is no fundamental difference from a p2p connection and bus connection, and both could serve a single interface or multiple interface/multiplexed communication. Imho, we shouldn't behave differently because of that.

> Also we don't need to provide an option to change this behaviour unless someone asks for it. I don't see much point in parallelising method calls from the same client. We should document this behaviour though.

There is as much reasons to want or no want parallelism with p2p or bus connection.

Reply:
> > There is no fundamental difference from a p2p connection and bus connection, and both could serve a single interface or multiple interface/multiplexed communication.
> 
> It's not about the number of interfaces served. It's about the relationship between client and connection 

The same p2p connection can be shared for different usages & interfaces, just like a bus connection.

> 
> > Imho, we shouldn't behave differently because of that.
> 
> The only use-case you presented was about p2p and if we can satisfy that use case w/o needing new API (at least temporarily), I'd prefer that.

It's not specific to p2p. org.qemu.Display1.Keyboard has Press and Release methods for example. There is no reason a client would have to wait for a reply for Press before sending Release. This interface is exposed on the default bus by default. The implementation is currently C/GDBus, but you should get my point.

Reply:
> > > There is no fundamental difference from a p2p connection and bus connection, and both could serve a single interface or multiple interface/multiplexed communication.
> > 
> > 
> > It's not about the number of interfaces served. It's about the relationship between client and connection
> 
> The same p2p connection can be shared for different usages & interfaces, just like a bus connection

As I wrote, it's not about interfaces (only if we take the approach in this PR, which I'm objecting to for other reasons I mentioned). 

> It's not specific to p2p. org.qemu.Display1.Keyboard has Press and Release methods for example. 

Again, dependencies between methods is **not** an argument alone. You can find countless examples of methods depending on each other out in the wild.

> There is no reason a client would have to wait for a reply for Press before sending Release.

If it's from the same client, yes they shouldn't need to. Ideally both in case of p2p and bus, but as I wrote, it'd be hard to implement it for the bus but in case of p2p, there is only one client per connection so it's easy to make it happen.

Reply:
> > The same p2p connection can be shared for different usages & interfaces, just like a bus connection
> 
> As I wrote, it's not about interfaces (only if we take the approach in this PR, which I'm objecting to for other reasons I mentioned).

I disagree, typically a set of related methods on an interface must be handled in order. If you are lucky enough to have an interface that doesn't need ordering, you can keep zbus default behaviour.

p2p or bus connection, it doesn't matter, they can serve several unrelated objects & interfaces with different constraints. Your point about making several p2p connections doesn't hold, they are usually sharing the same connection to uniquely identify a client and avoid pointless extra resources usage and connection setups.

> 
> > It's not specific to p2p. org.qemu.Display1.Keyboard has Press and Release methods for example.
> 
> Again, dependencies between methods is **not** an argument alone. You can find countless examples of methods depending on each other out in the wild.

Good you agree they are countless existing examples where ordering or methods matters!

Reply:
> Good you agree they are countless existing examples where ordering of methods matters! (the norm, not the exception to me)

I never disagreed with that at all. If you think I did, you misunderstood. A method call in general is incomplete w/o a return. If you had 2 async methods with interdependency, you won't do something like this in Rust:

```rust
// `method_b` depends on the side effects of `method_a` here.
let batch = [method_a(), method_b()];
...
for method in batch {
    tokio::task::spawn(method);
}
```

---

Since it's clear to me that we won't be able to convince each other after having spent a lot of our time on this already, I am no longer interested in discussing it further. I hate to do this but I'm pulling the maintainer card and ask you to either take one of the 2 options I suggested (either make non-spawning default only for the p2p and document this behavior, or add the ConnectionBuilder/ObjectServer method to disable task spawning) or close the PR.

Reply:
With task spawning, it's quite explicit that there are no dependencies. The trouble is the client is simply doing `send([methodA,methodB])` and can reasonably expect that to be handled in order, unless the implementation clearly state otherwise.

Reply:
Setting a no-spawn option _globally_ on Connection/ObjectServer brings back the potential dead-lock issue https://github.com/dbus2/zbus/commit/2da0ecd98f449aa2fcae2ff7f656fe91a13d8b9a. By setting `no_spawn` on the Interface, the _specific_ implementation states that it doesn't rely on reentering (no blocking calls etc).

Reply:
> Setting a no-spawn option _globally_ on Connection/ObjectServer brings back the potential dead-lock issue [2da0ecd](https://github.com/dbus2/zbus/commit/2da0ecd98f449aa2fcae2ff7f656fe91a13d8b9a)

I don't think so. The commit was made before ObjectServer started using its own MessageStream (i-e queue), which is separate from the stream used by method replies so the deadlock mentioned in the commit log is not possible anymore. Almost everything uses its own stream now in zbus even and this was done to avoid all such deadlocks. 

But maybe I'm missing something. Can describe where and how this can deadlock?

Reply:
The stream is still bounded (by `DEFAULT_MAX_QUEUED`). Since it's bounded, you could get this deadlock if you wait on any dbus recv operation inside your handler (the usual cause of such a wait would be "making a dbus call").  And even if it were unbounded (or the bound is not hit), you can still get a deadlock if you have a circular dependency of dbus calls that cycles back to this ObjectServer.  The circular dependency is a bit easier to just document as "don't risk doing that if you set no-spawn".

Reply:
@danieldg +1, correct

Reply:
> Since it's bounded, you could get this deadlock if you wait on any dbus recv operation inside your handler (the usual cause of such a wait would be "making a dbus call").

True. However, although after all the debates on this topic, I still remain unconvinced that anyone other than @elmarco needs this anyway but just in case I'm wrong and there are other people/projects out there "batching method calls with inter-dependencies", I am open to a connection option and I think it's more than sufficient to warn about this possible deadlock in the docs.

> you can still get a deadlock if you have a circular dependency of dbus calls that cycles back to this ObjectServer.

D-Bus calls are always more expensive and have a higher latency so using it for intra-process communication sounds like a bad idea in general. There are much better ways of doing that, e.g channels. So as you said, we shouldn't need to care about this case at all. Just document it, if really needed.

Reply:
> > Since it's bounded, you could get this deadlock if you wait on any dbus recv operation inside your handler (the usual cause of such a wait would be "making a dbus call").

Actually looking back at the code, I'm not sure that's true. The ObjectServer's receiver is [for `METHOD_CALL` messages only](https://github.com/dbus2/zbus/blob/f45718536c7fba6b88c3db071bb90ccc2efa4522/zbus/src/connection/mod.rs#L957) so [any `METHDO_RETURN` messages won't be broadcasted to it](https://github.com/dbus2/zbus/blob/6a78d8ef191391c69c4fffe530741db67a314bac/zbus/src/connection/socket_reader.rs#L63) and hence socket reader task won't deadlock on broadcasting any method call returns while the method dispatch is waiting on a method call.

Did I miss something?

Reply:
Yes: if parties on the bus have sent us `DEFAULT_MAX_QUEUED + 1` method calls, then `receive_msg` will block on the `broadcast_direct` call.  That prevents any new messages, regardless of type, from being processed.

Reply:
> Yes: if parties on the bus have sent us `DEFAULT_MAX_QUEUED + 1` method calls, then `receive_msg` will block on the `broadcast_direct` call.

But why? The queues are separate. Perhaps it would help me understand if you could describe an example scenario.

Reply:
1. Client A makes the AccidentalDeadlockFunction method call; this ends in zbus's queue.
2. Client B sends `DEFAULT_MAX_QUEUED + 1` method calls to `GetFoo` (say, a badly designed API where you can get only one property at a time; this doesn't matter exactly, just the number of messages).
3. https://github.com/dbus2/zbus/blob/f45718536c7fba6b88c3db071bb90ccc2efa4522/zbus/src/connection/socket_reader.rs#L75 starts sending to the queue.  After `DEFAULT_MAX_QUEUED`, it blocks (2 messages left to send).
4. The ObjectServer dequeues one method call (ADF) and calls it. It makes a call to `DBusProxy::GetConnectionUnixProcessId`; this gets sent out.
5. The dbus server sends the reply on the socket.
6. `receive_msg` runs again, and is able to send one more message via `broadcast_direct`.  There's one `GetFoo` left (parsed but not sent), and just behind it is the reply from `GetConnectionUnixProcessId`.
7. Deadlock.

Reply:
Thanks for explaining! Yeah, now I see it.

However, does having the option on the interface level eliminate this deadlock possibility? I think it only reduces the likelihood, assuming that developer disabling spawning, designs their interfaces and code carefully. If so, I don't see this as a good rationale for putting it on the interface level.

Reply:
To avoid the deadlock, you could have the documentation declare: if you ever wait on this `Connection` object inside your `ObjectServer` (by making a method call without NoReply, etc), you must declare your interface to be spawning (or if the default is spawning, must not declare it no-spawn).  The `GetFoo` calls in this example could be declared no-spawn without any problems.

Putting it at the interface level is "closer" to the problem (you could even put it at the method level) and so it's more likely to be noticed - though you could argue that either way.  If you have *multiple* interfaces, making the entire connection no-spawn puts restrictions on all of them - which is more of a burden to check.

Reply:
> Putting it at the interface level is "closer" to the problem (you could even put it at the method level)

Well, I think if we do this in a more granular way, method-level seems like the best option. If you hit this issue on an interface-level, there may not be something you could do about it (e.g if you're implementing a standardized interface, that you can not change).

@elmarco can you modify the PR to move the option to methods instead? I know this will require breaking changes on `Interface` but [as I wrote](https://github.com/dbus2/zbus/issues/674#issuecomment-2020310501) in #674, I think we should treat this an unstable API. We just need to document it more explicitly.

If so, please address other comments too and rebase the PR.

Reply:
> I think we should treat this an unstable API. We just need to document it more explicitly.

Having said that, we should still try to make the change a bit more future-proof:

```rust

pub struct CallSettings {
    spawn_task: bool,
}

impl CallSettings {
    pub fn new() {
        MethodCallSettings::default()
    }

    pub fn set_spawn_task(mut self, spawn_task: bool) -> Self {
         self.spawn_task = spawn_task;

         self
    }
}
```

and then add an `Option<CallSettings>` to `Interface::{call, call_mut}`.

Reply:
@zeenix if we are breaking Interface, I would rather implement part of my proposal from #674. Property and Method can have additional details, and there is no call & call_mut API duplication.

To me, the method granularity isn't the most appropriate. Usually, you have a related methods in an interface that you want to handle in order. Annotating each interface method is cumbersome, but that's fine to me if you believe it's better.